layer_factory/assistant_layer/mem_bound_warning.cpp

```diff
@@ -0,0 +1,210 @@
+VkResult MemBoundWarning::PostCallBindImageMemory(VkDevice device, VkImage image, VkDeviceMemory memory, VkDeviceSize memoryOffset,
+                                                  VkResult result) {
+    if (result == VK_SUCCESS) SetBindMemoryState<VkImage>(image, memory);
+
+    return VK_SUCCESS;
+}
+
+VkResult MemBoundWarning::PostCallBindImageMemory2(VkDevice device, uint32_t bindInfoCount,
+                                                   const VkBindImageMemoryInfoKHR *pBindInfos, VkResult result) {
+    if (result == VK_SUCCESS) {
+        for (uint32_t i = 0; i < bindInfoCount; i++) {
+            SetBindMemoryState<VkImage>(pBindInfos[i].image, pBindInfos[i].memory);
+        }
+    }
+
+    return VK_SUCCESS;
+}
+
+VkResult MemBoundWarning::PostCallBindImageMemory2KHR(VkDevice device, uint32_t bindInfoCount,
+                                                      const VkBindImageMemoryInfo *pBindInfos, VkResult result) {
+    if (result == VK_SUCCESS) {
+        for (uint32_t i = 0; i < bindInfoCount; i++) {
+            SetBindMemoryState<VkImage>(pBindInfos[i].image, pBindInfos[i].memory);
+        }
+    }
+
+    return VK_SUCCESS;
+}
+
+// intercept for the sparse memory to track memory state
+VkResult MemBoundWarning::PostCallQueueBindSparse(VkQueue queue, uint32_t bindInfoCount, const VkBindSparseInfo *pBindInfo,
+                                                  VkFence fence, VkResult result) {
+    for (uint32_t bindIdx = 0; bindIdx < bindInfoCount; bindIdx++) {
+        const VkBindSparseInfo &bindInfo = pBindInfo[bindIdx];
+
+        for (uint32_t j = 0; j < bindInfo.bufferBindCount; j++) {
+            for (uint32_t k = 0; k < bindInfo.pBufferBinds[j].bindCount; k++) {
+                auto sparse_binding = bindInfo.pBufferBinds[j].pBinds[k];
+                SetBindMemoryState<VkBuffer>(bindInfo.pBufferBinds[j].buffer, sparse_binding.memory);
+            }
+        }
+        for (uint32_t j = 0; j < bindInfo.imageOpaqueBindCount; j++) {
+            for (uint32_t k = 0; k < bindInfo.pImageOpaqueBinds[j].bindCount; k++) {
+                auto sparse_binding = bindInfo.pImageOpaqueBinds[j].pBinds[k];
+                SetBindMemoryState<VkImage>(bindInfo.pImageOpaqueBinds[j].image, sparse_binding.memory);
+            }
+        }
+        for (uint32_t j = 0; j < bindInfo.imageBindCount; j++) {
+            for (uint32_t k = 0; k < bindInfo.pImageBinds[j].bindCount; k++) {
+                auto sparse_binding = bindInfo.pImageBinds[j].pBinds[k];
+                SetBindMemoryState<VkImage>(bindInfo.pImageBinds[j].image, sparse_binding.memory);
+            }
+        }
+    }
+
+    return VK_SUCCESS;
+}
+
+VkResult MemBoundWarning::PostCallGetSwapchainImagesKHR(VkDevice device, VkSwapchainKHR swapchain, uint32_t *pSwapchainImageCount,
+                                                        VkImage *pSwapchainImages, VkResult result) {
+    assert(swapchain);
+    auto images = swapchain_image_map[swapchain];
```

Comment:
It looks like this is making a copy. Did you intend to use a reference instead?

Reply:
Fixed in upcoming commit